### PR TITLE
Fix stale product term caches after reordering

### LINKED ISSUE #67709: Reordering attribute terms never invalidates the per-product terms cache; custom variation dropdown order silently ignored on the storefront until an unrelated product save
## TL;DR

`wc_set_term_order()` writes the `order` term meta but never invalidates the `product_<id>` cache group that `_wc_get_cached_product_terms()` writes to. On any host with a persistent object cache, dragging attribute terms into a custom order updates the admin list and does nothing on the storefront. There is no TTL — it never self-corrects with time, only when something unrelated happens to save the product. Confirmed by A-B-A control test.

---

## Describe the bug

A merchant sets a custom order for a global attribute's terms (Products → Attributes → Size → Configure terms → drag). The admin term list immediately shows the new order. The variation dropdown on the product page does not, and will not — no action on the attribute screen clears the cache serving the old order, and the cached entry has no expiry. The only fix is saving the product, which is unrelated, undocumented, and undiscoverable.

The reasonable merchant conclusion is that attribute term ordering is broken. In the case that surfaced this, that's exactly what happened.

## Steps to reproduce

0. **Requires a persistent object cache (Redis/Memcached).** Not reproducible without one — see Technical context.
1. Create global attribute `pa_size` with Default sort order = **Custom ordering**, terms Small, Medium, Large, Extra Large.
2. Create a variable product using `pa_size`, generate variations, price them. Leave stock management off (see Impact).
3. View the product page (or run the first command below) to warm the cache. Order is alphabetical — correct, since no `order` meta is set yet.
4. Products → Attributes → Size → Configure terms → drag into Small, Medium, Large, Extra Large. Confirm the admin list reflects it.
5. Verify via WP-CLI, which keeps page caching out of the picture entirely:

```
wp eval 'print_r( wp_list_pluck( wc_get_product_terms( PRODUCT_ID, "pa_size", array( "fields" => "all" ) ), "name" ) );'
```

```
wp eval 'print_r( wp_list_pluck( get_terms( array( "taxonomy" => "pa_size", "hide_empty" => false ) ), "name" ) );'
```

6. **Expected:** both return `Small, Medium, Large, Extra Large`.
7. **Actual:** the first returns the stale `Extra Large, Large, Medium, Small`; the second returns the correct order. The storefront dropdown matches the stale one.
8. Wait any length of time and re-run — still stale. Re-save the product → both agree, storefront corrects itself.

## Verified reproduction (JN site)

Reproduced on a clean site: **WordPress 7.0.2, WooCommerce 10.9.4, Memcache object cache.** Every step driven through WP-CLI, so page caching is not involved at any point.

A-B-A control, toggling only the `object-cache.php` drop-in on one host (so hosting cannot explain the difference):

| Object cache | After reorder, no product save | Result |
| -- | -- | -- |
| Memcache (as shipped) | `wc_get_product_terms` → `Small | Medium | Large | Extra Large` · `get_terms` → `Extra Large | Large | Medium | Small` | **stale** |
| Drop-in removed (`Default`) | both paths → `Extra Large | Large | Medium | Small` | correct, immediately |
| Memcache restored | `wc_get_product_terms` → `Small | Medium | Large | Extra Large` · `get_terms` → `Extra Large | Large | Medium | Small` | **stale again** |

`order` term meta was written correctly in all three phases — the write path is fine, only the read is stale.

End-to-end through `wc_dropdown_variation_attribute_options()`: rendered `Extra Large | Large | Medium | Small` before re-saving the product, `Small | Medium | Large | Extra Large` after.

## Technical context (verified against `trunk` @ `66a25f33c7`, 11.1.0-dev; reproduced on 10.9.4)

`_wc_get_cached_product_terms()` — `includes/wc-term-functions.php:154-168` — caches the ordered term list per product with no expiry:

```
$cache_group = WC_Cache_Helper::get_cache_prefix( 'product_' . $product_id ) . $product_id;  // :156
$terms       = wp_cache_get( $cache_key, $cache_group );                                      // :157
if ( false !== $terms ) { return $terms; }                                                    // :159-161
...
wp_cache_add( $cache_key, $terms, $cache_group );                                             // :165
```

This is the only read path for dropdown ordering, reached via `wc_get_product_terms()` (`:178`). Both renderers use it, so block vs classic makes no difference:

* Classic — `includes/wc-template-functions.php:3762`
* Blocks — `src/Blocks/BlockTypes/AddToCartWithOptions/VariationSelectorAttribute.php:204`

`wc_set_term_order()` — `includes/wc-term-functions.php:363-384` — what a drag reaches via `WC_AJAX::term_ordering()` (`includes/class-wc-ajax.php:2310`) → `wc_reorder_terms()` (`:305`):

```
update_term_meta( $term_id, 'order', $index );   // :368
if ( ! $recursive ) { return $index; }           // :370-372
...
clean_term_cache( $term_id, $taxonomy );         // :381
```

It writes the meta and never invalidates the `product_<id>` group. The only callers of `WC_Cache_Helper::invalidate_cache_group( 'product_…' )` anywhere in the codebase are in `WC_Product_Data_Store_CPT::clear_caches()` — `includes/data-stores/class-wc-product-data-store-cpt.php:1160-1167` — which runs only on product create (`:258`) and update (`:389`). Nothing on the term-ordering path reaches it.

`woocommerce_after_set_term_order` fires at `:337` but has **no listeners in core** — an available hook, not a working invalidation.

### The stale entry has no TTL

`wp_cache_add()` at `:165` passes no expiry. On WordPress.com/Atomic the platform drop-in routes every write through `get_expiration()`, which falls back to `default_expiration` when the requested expiry is `<= 0` — and that default is `0`, i.e. memcached "never expires" (`/scripts/object-cache.memcached.php:432-433, :1505-1512`; platform code, quoted for context, not WooCommerce). So the entry persists until explicitly invalidated or evicted by LRU under memory pressure.

**What does clear it:** any save of the product or one of its variations, since `clear_caches()` invalidates the parent's group too for variations (`:1162-1164`). In practice: the merchant pressing Update, bulk/quick edit, CSV import or feed sync, a scheduled sale price transition, a refund restock — or a sale.

A sale clears it because `wc_update_product_stock()` saves the product "so stock status etc is in sync, and caches are cleared" (`includes/wc-stock-functions.php:62-64`). Verified on the repro site: reorder terms without touching the product, then decrement stock via `wc_update_product_stock()`, and both read paths agree again.

**But that incidental healing has two holes that align badly with when merchants hit this.** Stock reduction is gated on stock management being enabled — `wc_reduce_stock_levels()` skips any item where `! $product->managing_stock()` (`includes/wc-stock-functions.php:194-195`) — and apparel/print-on-demand stores commonly run without it. More importantly, merchants configure size ordering *during product setup*, before the product has ever sold, so there is no incidental save coming at all. The population most likely to trigger the bug is the population least likely to be rescued from it.

### Why a persistent object cache is required

`get_cache_prefix()` stores the namespace via `wp_cache_get`/`wp_cache_add` (`src/Caching/CacheNameSpaceTrait.php:23`), and `generate_cache_prefix()` is `microtime() . random_bytes(8)` (`:105-107`). With no persistent backend every request misses, mints a fresh prefix, and lands in a different cache group — so a stale entry can never be read back. This is almost certainly why the bug has survived: it is invisible in local dev and in any environment without an object cache.

### Secondary observations (not the root cause)

* `clean_term_cache()` at `:381` is unreachable for non-recursive calls because of the early return at `:370-372`. `wc_reorder_terms()` calls non-recursively for every term except the moved one (`:332` vs `:327`), so most calls skip it.
* The admin list *does* refresh, indicating WP's own term-query caches are invalidated by the meta write. Only the WooCommerce per-product wrapper goes stale. That admin/storefront disagreement is what makes this so confusing in the field.
* Ordering itself is correct once the cache is bypassed: `wc_terms_clauses()` sorts `meta_value+0 ASC, t.name` (`:115`), so unset `order` meta ties at 0 and falls back to alphabetical — the pre-drag behavior.

## Impact

**Who is affected:** any merchant using global attributes with Custom ordering **on a host with a persistent object cache** — Atomic/WordPress.com, Pressable, WP Engine, Kinsta, or self-hosted with Redis/Memcached. That covers most production stores above hobby scale. Apparel and sized goods are the obvious blast radius, since correct size ordering is table stakes there.

**Duration:** indefinite. No TTL, no background refresh. Cleared only by an unrelated product save, and the two most common sources of such a save — merchant edits and stock-decrementing sales — are both absent during initial product setup, which is exactly when merchants set term order. Stores with stock management disabled never get the sale-triggered clear at all.

**Workaround:** re-save the product (bumps `product_<id>` via `clear_caches()`), or flush the object cache. Must be repeated per product using the attribute. Not discoverable without reading the source.

## Suggested fix

Invalidate the `product_<id>` cache group for products carrying the reordered terms. `woocommerce_after_set_term_order` (`includes/wc-term-functions.php:337`) already passes term, index, and taxonomy, so the hook point exists.

One implementation caveat: that action fires once per term inside `wc_reorder_terms()`'s loop over the whole level, so hooking it naively means a product lookup per term. Invalidating once after the reorder completes — or collecting affected product IDs and invalidating in bulk — is the better shape, since this runs inside an AJAX request on every drag and attribute term sets can be large.

Worth checking whether the early return at `:370-372` skipping `clean_term_cache()` is intentional while in there.

---

Investigated and written with Claude Code. 

First seen in 11570789-zen while helping Atomic user, then reproduced on JN site.

### PR DESCRIPTION
**Pull request by gigitux**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR fixes stale `wc_get_product_terms()` results after product taxonomy terms are reordered.

This PR adds a taxonomy-scoped cache namespace and invalidates it whenever WordPress cleans a taxonomy associated with products. This covers both AJAX reordering and direct `wc_reorder_terms()` calls.

Closes #67709.

Closes #67709.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://developer.woocommerce.com/docs/contribution/testing/writing-high-quality-testing-instructions/), include your detailed testing instructions:

0. **Requires a persistent object cache (Redis/Memcached).** Not reproducible without one — see Technical context.
1. Create global attribute `pa_size` with Default sort order = **Custom ordering**, terms Small, Medium, Large, Extra Large.
2. Create a variable product using `pa_size`, generate variations, price them. Leave stock management off (see Impact).
3. View the product page (or run the first command below) to warm the cache. Order is alphabetical — correct, since no `order` meta is set yet.
4. Products → Attributes → Size → Configure terms → drag into Small, Medium, Large, Extra Large. Confirm the admin list reflects it.
5. Verify via WP-CLI, which keeps page caching out of the picture entirely:

```
wp eval 'print_r( wp_list_pluck( wc_get_product_terms( PRODUCT_ID, "pa_size", array( "fields" => "all" ) ), "name" ) );'
```

```
wp eval 'print_r( wp_list_pluck( get_terms( array( "taxonomy" => "pa_size", "hide_empty" => false ) ), "name" ) );'
```

6. **Expected:** both return `Small, Medium, Large, Extra Large`.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

Created manually.

</details>

### Release Communication

Select the release communication labels this PR needs:

- [ ] **Feature Highlight** - For user-facing features (what changed, user impact)
- [ ] **Developer Advisory** - For developer-facing changes (what changed, how to detect, actions needed)

<details>
<summary>When to use each?</summary>

**Feature Highlight**: New features, UI changes, or improvements that merchants/store owners will notice.

- Example: "New bulk editing for products", "Improved checkout performance"

**Developer Advisory**: Breaking changes, deprecations, or changes that affect themes/plugins/extensions.

- Example: "Hook signature change", "Deprecated filter", "REST API field removed"

Selected labels are applied when the pull request is merged.
</details>

### Use of AI Tools

Codex was used to investigate the cache behavior, implement the change and regression test, run static verification, and draft the PR description. The resulting changes were reviewed during development.

### CODE REVIEW

**Comment by gigitux** [comment]
@coderabbitai review

**Review by dinhtungdu — CHANGES_REQUESTED** [review]
I don't think this got a new service/class. We already have a method hooked into `clean_term_cache`, should we update that instead?

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-cache-helper.php#L315

**Comment by gigitux** [comment]
> I don't think this got a new service/class. We already have a method hooked into `clean_term_cache`, should we update that instead?
> 
> https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-cache-helper.php#L315

Good question, and thanks for raising it, I hadn’t considered reusing the existing method.
I’m open to moving the logic there, but before doing it, I think that there are some pros with the current approach:

- `WC_Cache_Helper` is a public class - do we want to add this logic to a public function? To be honest, I feel that the cache logic should be "private" and just exposed based on the feedback. In my mind, WooCommerce should be able to cover most of the cases when the cache should be cleaned up. I know that third-party extensions can still copy and paste code, but having logic private may help to prevent abuse. 
- I find a dedicated internal class for each cache domain easier to understand and trace. It makes the invalidation flow and its responsibility more explicit.

**Review by dinhtungdu — APPROVED** [review]
> I find a dedicated internal class for each cache domain easier to understand and trace. It makes the invalidation flow and its responsibility more explicit.

This makes sense. Let's go with a new service.

Note that currently, the invalidation happens before the reordering finishes, so we may have partial ordering on busy stores with many attribute terms. I considered this an edge case so I don't think we need to fix it, at least for now without any report. I will let you decide, hence approving here.

Also can you update the PR testing instruction, currently it's the same as issue's step to reproduce?

**Comment by gigitux** [comment]
> Note that currently, the invalidation happens before the reordering finishes, so we may have partial ordering on busy stores with many attribute terms. I considered this an edge case so I don't think we need to fix it, at least for now without any report. I will let you decide, hence approving here.

I'm planning to merge this PR and see if we receive some bug reports to iterate on it. 

> Also can you update the PR testing instruction, currently it's the same as issue's step to reproduce?

Yes, these ones are mentioned in #67709
